casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -35,54 +37,88 @@ object EstimatorHelper {
     } yield result
   }
 
-  /*
-   * Block b1 conflicts with b2 if any of b1's ancestors contains a replay log entry that
-   * touches a channel that any of b2's ancestors' (that are not common with b1's ancestors)
-   * replay log entries touch.
-   */
   private[casper] def conflicts[F[_]: Monad: Log: BlockStore](
       b1: BlockMessage,
       b2: BlockMessage,
       dag: BlockDagRepresentation[F]
   ): F[Boolean] =
-    dag.deriveOrdering(0L).flatMap { // TODO: Replace with something meaningful
-      implicit ordering =>
-        for {
-          b1MetaDataOpt        <- dag.lookup(b1.blockHash)
-          b2MetaDataOpt        <- dag.lookup(b2.blockHash)
-          blockMetaDataSeq     = Vector(b1MetaDataOpt.get, b2MetaDataOpt.get)
-          uncommonAncestorsMap <- DagOperations.uncommonAncestors[F](blockMetaDataSeq, dag)
-          (b1AncestorsMap, b2AncestorsMap) = uncommonAncestorsMap.partition {
-            case (_, bitSet) => bitSet == BitSet(0)
-          }
-          b1AncestorsMeta    = b1AncestorsMap.keys
-          b2AncestorsMeta    = b2AncestorsMap.keys
-          b1AncestorChannels <- buildBlockAncestorChannels[F](b1AncestorsMeta.toList)
-          b2AncestorChannels <- buildBlockAncestorChannels[F](b2AncestorsMeta.toList)
-          conflicts          = b1AncestorChannels.intersect(b2AncestorChannels).nonEmpty
-          _ <- if (conflicts) {
-                Log[F].info(
-                  s"Block ${PrettyPrinter.buildString(b1.blockHash)} and ${PrettyPrinter
-                    .buildString(b2.blockHash)} conflicts."
-                )
-              } else {
-                Log[F].info(
-                  s"Block ${PrettyPrinter
-                    .buildString(b1.blockHash)}'s channels ${b1AncestorChannels.map(PrettyPrinter.buildString).mkString(",")} and block ${PrettyPrinter
-                    .buildString(b2.blockHash)}'s channels ${b2AncestorChannels.map(PrettyPrinter.buildString).mkString(",")} don't intersect."
-                )
-              }
-        } yield conflicts
+    dag.deriveOrdering(0L).flatMap { implicit ordering =>
+      for {
+        b1MetaDataOpt <- dag.lookup(b1.blockHash)
+        b2MetaDataOpt <- dag.lookup(b2.blockHash)
+        uncommonAncestorsMap <- DagOperations.uncommonAncestors[F](
+                                 Vector(b1MetaDataOpt.get, b2MetaDataOpt.get),
+                                 dag
+                               )
+        (b1AncestorsMap, b2AncestorsMap) = uncommonAncestorsMap.partition {
+          case (_, bitSet) => bitSet == BitSet(0)
+        }
+        b1Events <- extractBlockEvents[F](b1AncestorsMap.keys.toList)
+        b2Events <- extractBlockEvents[F](b2AncestorsMap.keys.toList)
+        conflictsBecauseOfJoins = extractJoinedChannels(b1Events)
+          .intersect(allChannels(b2Events))
+          .nonEmpty || extractJoinedChannels(b2Events).intersect(allChannels(b1Events)).nonEmpty
```

Comment:
it'd make sense to reuse the keys of `tuplespaceEventsPerChannel(b1Events)` instead of computing the allChannels(b1Events) separately

Reply:
This means we'd have to always construct the `tuplespaceEventsPerChannel` map, which is costly. I'd leave this as is, because currently if two blocks conflict on joined channels, we fail fast without constructing it